pkg/graveler/committed/merge.go

```diff
@@ -0,0 +1,308 @@
+package committed
+
+import (
+	"bytes"
+	"context"
+
+	"github.com/treeverse/lakefs/pkg/graveler"
+)
+
+type mergeState int
+
+const (
+	mergeStateInvalid mergeState = iota
+	// Range states
+	mergeStateInitial
+	mergeStateRangeAddedBySource
+	mergeStateRangeAddedByDest
+	mergeStateRangeNoChange
+	mergeStateRangeRemovedBySource
+	mergeStateRangeRemovedByDest
+	mergeStateRangeConflict
+	mergeStateValueAddedBySource
+	mergeStateValueAddedByDest
+	mergeStateValueRemovedBySource
+	mergeStateValueRemovedByDest
+	mergeStateValueChangedBySource
+	mergeStateValueChangedByDest
+	mergeStateValueNoChange
+	mergeStateValueConflicting
+
+	mergeStateEnterSourceRange
+	mergeStateEnterDestRange
+)
+
+type mergeSt struct {
+	baseIter, sourceIter, destIter    Iterator
+	baseNext, sourceNext, destNext    bool
+	baseValue, sourceValue, destValue *graveler.ValueRecord
+	baseRange, sourceRange, destRange *Range
+}
+
+func Merge(ctx context.Context, writer MetaRangeWriter, baseIter Iterator, sourceIter Iterator, destIter Iterator, opts *ApplyOptions) (graveler.DiffSummary, error) {
+	ret := graveler.DiffSummary{Count: make(map[graveler.DiffType]int)}
+	var sourceValue, destValue *graveler.ValueRecord
+	var sourceRange, destRange *Range
+	sourceNext, destNext := sourceIter.Next(), destIter.Next()
+	for sourceNext && destNext {
+		select {
+		case <-ctx.Done():
+			return graveler.DiffSummary{}, ctx.Err()
+		default:
+		}
+		sourceValue, sourceRange = sourceIter.Value()
+		destValue, destRange = destIter.Value()
+		state, err := getState(baseIter, sourceValue, sourceRange, destValue, destRange)
+		if err != nil {
+			return graveler.DiffSummary{}, err
+		}
+		switch state {
+		case mergeStateRangeAddedBySource:
+			ret.Count[graveler.DiffTypeAdded] += int(sourceRange.Count)
+			sourceNext = sourceIter.NextRange()
+			err = writer.WriteRange(*sourceRange)
+		case mergeStateRangeAddedByDest:
+			destNext = destIter.NextRange()
+			err = writer.WriteRange(*destRange)
+		case mergeStateRangeNoChange:
+			destNext = destIter.NextRange()
```

Comment:
Shouldn't this also skip source?
```suggestion
			sourceNext = sourceIter.NextRange()
			destNext = destIter.NextRange()			
```